### PR TITLE
[DEM-736] flush after run triggers exception, type hints and unittests

### PR DESCRIPTION
**Pull request by QFer**
* Added member _flushed which is set after run() and triggers an exception in _store when a gate, other than deallocate, is stored.
* Added an exception when Api is None when constructing QIBackend, because it is mandatory to execute the qasm.
* _reset routine was resetting the allocated_qubits after _run(), but projectq is sending DeAllocate for the allocated qubits. Removed from _reset. In _store the Deallocate gate is handled now  as a reverse Allocate gate by removing the qubit out of the set and setting max_qubit_id to the max qubit of the set or -1 if empty. max_qubit_id is now really the maximum bitnumber in the set of allocated_qubits. When a Deallocate was handled, max_qubit_id wasn't correct.
* unused self._gate removed
* Type hints
* Added unittest for both added exceptions.
* Renamed an unittest that throws an exception but was name result_oke or something.

### CODE REVIEW

File: src/quantuminspire/projectq/backend_qx.py [hunk 1]
```diff
@@ -37,37 +39,45 @@ class QIBackend(BasicEngine):
...
+        any additional gates received after a FlushGate triggers an exception. """
         self._reset()
         self._num_runs = num_runs
         self._verbose = verbose
         self._cqasm = str()
         self._measured_states = {}
+        self._measured_ids = []
+        self._allocated_qubits = set()
+        self._max_qubit_id = -1
+        if not quantum_inspire_api:
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/projectq/backend_qx.py]: Exactly. I did add the check to fail fast.

File: src/quantuminspire/projectq/backend_qx.py [hunk 2]
```diff
@@ -37,37 +39,45 @@ class QIBackend(BasicEngine):
...
-            backend_type (dict or str or None): Backend to use for execution.
+            num_runs: Number of runs to collect statistics (default is 1024).
+            verbose: Verbosity level, defaults to 0, which produces no extra output.
+            quantum_inspire_api: Connection to QI platform, required parameter.
+            backend_type: Backend to use for execution. When backend_type = None, te default backend will be
+                          used.
         """
         BasicEngine.__init__(self)
+        self._flushed = False
+        """ bool: Because engines are meant to be 'single use' by the way ProjectQ is designed,
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/projectq/backend_qx.py]: In example google docstring they add docstring this way:
`        self.attr5 = None
        """str: Docstring *after* attribute, with type specified."""
'

File: src/quantuminspire/projectq/backend_qx.py [hunk 3]
```diff
@@ -247,15 +262,15 @@ def get_probabilities(self, qureg):
...
 
         Args:
-            state (int): state represented as an integer number
-            qureg (list<Qubit>): list of qubits for which to extract the state bit
+            state: state represented as an integer number
+            qureg: list of qubits for which to extract the state bit
 
         Returns:
-            (string): a string of '0' and '1' corresponding to the bit value in state of each Qubit in qureg
+            a string of '0' and '1' corresponding to the bit value in state of each Qubit in qureg
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/projectq/backend_qx.py]: I'll do a check for the whole file.

File: src/tests/quantuminspire/projectq/test_backend_qx.py [hunk 4]
```diff
@@ -237,6 +240,30 @@ def test_receive(self, function_mock):
...
+    @patch('quantuminspire.projectq.backend_qx.get_control_count')
+    def test_reuse_after_flush_RaisesRuntimeError(self, function_mock):
+        function_mock.return_value = 1
+        command = MagicMock(gate=NOT, qubits=[[MagicMock(id=0)], [MagicMock(id=1)]])
+        command_list = [command, MagicMock(gate=FlushGate()), command]
+        api = MockApiClient()
+        backend = QIBackend(quantum_inspire_api=api)
+        backend.main_engine = MagicMock()
+        with patch('sys.stdout', new_callable=io.StringIO):
+            self.assertRaises(RuntimeError, backend.receive, command_list)
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/tests/quantuminspire/projectq/test_backend_qx.py]: Ok

**Comment by QFer** [comment]
Fixed the grover example. Interpreted _max_qubit_id differrently. Added extra docstring. The set() `self._allocated_qubits` is superfluous imo. It is not really used.

**Comment by QFer** [comment]
Added test_maximum_qubit. 3 qubits (0..2) allocated, 3 deallocated. Tests that the set _allocated_qubits is empty but _number_of_qubits = 3

File: src/quantuminspire/projectq/backend_qx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (qSaevar) [reply src/quantuminspire/projectq/backend_qx.py]: But that is if you don´t use typehints, now we are typehinting twice.

**Comment by QFer** [comment]
I removed the last duplicate docstring type hint

**Comment by qSaevar** [comment]
One more thing, this branch is out of data. You need to rebase before I approve.

**Pull request merged**